packages/core/src/parser.ts

```diff
@@ -0,0 +1,155 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: @loopback/core
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {ServerRequest as Request, ServerResponse as Response} from 'http';
+import {Application} from './Application';
+import {invoke} from './invoke';
+import {SwaggerRouter} from './router/SwaggerRouter';
+import {Endpoint} from './router/SwaggerRouter';
+import {OpenApiSpec, OperationObject, ParameterObject} from '@loopback/openapi-spec';
+import * as assert from 'assert';
+const debug = require('debug')('loopback:request-parser');
+import * as Promise from 'bluebird';
+import * as url from 'url';
+import * as pathToRegexp from 'path-to-regexp';
+
+type jsonBodyFn = (req: Request, cb: (err?: Error, body?: {}) => void) => void;
+const jsonBody: jsonBodyFn = require('body/json');
+
+// tslint:disable:no-any
+type MaybeBody = any | undefined;
+export type OperationArgs = any[];
+type PathParameterValues = {[key: string]: any};
+type OperationRetval = any;
+// tslint:enable:no-any
+
+const parseJsonBody: (req: Request) => Promise<MaybeBody> = Promise.promisify(jsonBody);
+
+// As per the earlier design discussion, parser() was suppose to take request and swagger spec 
+// as arguments. However, this parsing/unmarshalling logic needs an instance of an Endpoint which is already created
+// in the router prior to parser(..) being called. Router would have mounted controllers and created endpoints
+// before parser() gets called. Hence Swagger spec is not needed here.
```

Comment:
IIUC, the only reason why we need the Endpoint instance is to access the arguments mapped to path parameters. IMO, this is something that should be handled by the router.

The signature of the parser function then can become something like the following:

```ts
export function parser(
  request: Request, pathParameters: StringToStringMap, spec: OperationObject)
  : OperationArgs
```